### PR TITLE
2 Factor Authentication improvements

### LINKED ISSUE #11177: Improve implementation for 2FA remember browser option
see also: https://github.com/friendica/friendica/issues/10690

### Currently

The wording is hard to understand and the places where management of trusted browsers is done in friendica is hard to remember and difficult to discover.

### Expected

The problem with current wording `This is my two-factor authenticator app device` is, that with a password manager which syncs with all devices, this is very confusing. Also OTPs are not only accessible on mobile but also on desktop when using password managers making this even more confusing. (the hint about recovery codes specifically mentions access to mobile.

Apple uses the text below which I think is a good solution and easy to grasp.

There is another smart aspect about how apple has implemented this browser trusting mechanism: They do not ask this question during login, but *after* the login has been sent:

```
Trust this browser?
If you choose to trust this browser, you will not be asked for a verification code the next time you sign in.

Options: Not Now | Don't Trust | Trust (defaulting to trust)
```

Screenshot 2022-01-23 at 12 56 18

And actually apple has done a good job with this implementation since when you log out you see another smart dialog:

```
Sign out of this browser?
If you trust this browser, you will not be asked for a verification code the next time you sign in.
Options: Trust and sign out | Cancel | Sign out (defaulting to Sign out)
```
Allowing to forget about the browser trusting on sign out, or keeping it and with that giving the user full control. That is really nicely done.

Screenshot 2022-01-23 at 13 04 32

The info about recovery codes is also unnecessarily confusing:
`Don’t have your phone? Enter a two-factor recovery code`
Proposal: `If you do not have access to your authentication code you can use a two-factor recovery code.`

### PR DESCRIPTION
**Pull request by nupplaphil**
Closes #11177

I struggled at the same point as @foss- that I don't fully understand, how the 2FA is **really** working.

I found the explanation at #11177 really self-explaining and adapted the 2FA to work like the same

- I renamed the wording for using the recovery codes
- I moved the checkbox "trusted_browser" during the verification into an own page
  - A new field `trusted` for each trusted_browser entry holds now if the trusted_browser should be trusted or not ;)
- If the trusted_browser is "trusted",
  - The logout page now redirects to `2fa/signout` asking the user, if he wants to remain the trust for this browser
  - If so, the next login skips the whole 2fa again
  - if not, the trusted_browser gets deleted as well
- if the trusted_browser is **not** trusted,
   - Never ask again about this browser, not at the next login, nor for any logout (you can only remove this browser per settings dialog)
   - **Every** login now prints the 2fa
- "Remember me" during the login doesn't have any effect on the 2fa behavior anymore

I tested every case above manually with my mobile device and webbrowser, so I'm pretty sure it should work ...

### CODE REVIEW

File: src/Module/Security/TwoFactor/Signout.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+class Signout extends BaseModule
+{
+	protected $errors = [];
+
+	/** @var IHandleSessions  */
+	protected $session;
+	/** @var Cookie  */
+	protected $cookie;
+	/** @var TwoFactor\Repository\TrustedBrowser  */
+	protected $trustedBrowserRepositoy;
```
Comment (MrPetovan) [inline src/Module/Security/TwoFactor/Signout.php]: Please fix this typo all across the file.
```suggestion
	protected $trustedBrowserRepository;
```

File: src/Module/Settings/TwoFactor/Index.php [hunk 2]
```diff
@@ -90,7 +91,9 @@ protected function post(array $request = [])
 					break;
 			}
 		} catch (\Exception $e) {
-			notice(DI::l10n()->t('Wrong Password'));
+			if (!($e instanceof FoundException)) {
+				notice(DI::l10n()->t($e->getMessage()));
+			}
 		}
```
Comment (MrPetovan) [inline src/Module/Settings/TwoFactor/Index.php]:
```suggestion
			notice(DI::l10n()->t($e->getMessage()));
		} catch (FoundException $e) {
			// Nothing to see here, please disperse
		}
```
I'm not sure about the order of the `catch` calls.

File: src/Module/Security/TwoFactor/Signout.php (hunk 1 above)
Comment (nupplaphil) [reply src/Module/Security/TwoFactor/Signout.php]: So You say you never heard of a reposiToy? :-D

**Comment by annando** [comment]
Out of interest: How are API calls treated? We allow all API calls to be either authenticated via OAuth or via HTTP Auth. OAuth should be fine in a 2FA environment, but the HTTP Auth for me seems to be huge problem here.

**Comment by tobiasd** [comment]
You can add passwords per app that are allowed to be used without the 2FA key for the API.

**Review by MrPetovan — APPROVED** [review]

**Comment by tobiasd** [comment]
Is it possible, that the re-worked 2FA disabled the workflow of the `startpage` addon? It is supposed to change the page shown after a login, roundabout since this commit it does not work on the initial login.

**Comment by MrPetovan** [comment]
Most probably.

**Pull request merged**